Note: This has been taken from https://engineering.ucsb.edu/~shell/che210d/numpy.pdf
(a list of useful numpy apis)

In [141]:
import numpy as np

### Array mathematics

In [142]:
a = np.array([1,2,3], float)
b = np.array([5,2,6], float)

In [143]:
a + b

array([ 6.,  4.,  9.])

In [144]:
a/b

array([ 0.2,  1. ,  0.5])

In [145]:
a % b

array([ 1.,  0.,  3.])

In [146]:
b % a

array([ 0.,  0.,  0.])

In [147]:
b**a #power

array([   5.,    4.,  216.])

In [140]:
a = np.array([1,2,3], float)
b = np.array([4,5], float)
a + b

ValueError: operands could not be broadcast together with shapes (3,) (2,) 

In [148]:
a = np.array([[1, 2], [3, 4], [5, 6]], float)
b = np.array([1, 100], float)
a + b

array([[   2.,  102.],
       [   4.,  104.],
       [   6.,  106.]])

In [149]:
np.array([-1, 3], float)

array([-1.,  3.])

In [150]:
b = np.array([-1., 3.], float)
b

array([-1.,  3.])

In [151]:
 b[:,np.newaxis]

array([[-1.],
       [ 3.]])

In [152]:
 b[:,]

array([-1.,  3.])

In [153]:
b[:,]

array([-1.,  3.])

In [154]:
x = b[:,np.newaxis]
x

array([[-1.],
       [ 3.]])

In [155]:
x.shape

(2, 1)

In [156]:
 b[np.newaxis,:]

array([[-1.,  3.]])

In [157]:
arr = np.arange(4)
print arr.shape
arr

(4,)


array([0, 1, 2, 3])

In [158]:
arr = np.arange(4)
arr[np.newaxis,:]

array([[0, 1, 2, 3]])

In [159]:
col_vec = arr[:, np.newaxis]

In [160]:
col_vec.shape

(4, 1)

In [161]:
a = np.array([[0, 2], [3, -1], [3, 5]], float)
a

array([[ 0.,  2.],
       [ 3., -1.],
       [ 3.,  5.]])

In [162]:
for (x, y) in a:
    print x+y

2.0
2.0
8.0


In [163]:
#equivalent but much much faster
a.sum(axis=1)

array([ 2.,  2.,  8.])

### Basic array operations

In [164]:
a = np.array([2, 4, 3], float)
print a.sum()
print a.prod()

9.0
24.0


In [165]:
a.std()

0.81649658092772603

In [166]:
np.sqrt(((a - a.mean())**2).sum()/a.size)

0.81649658092772603

In [167]:
#The argmin and argmax functions return the array indices of the minimum and maximum values:
a = np.array([10, 100, 0, 9999], float)
print a.argmax()
print a.argmin()

3
2


In [168]:
a = np.array([[0, 200], [3, -1], [3, 5]], float)
a

array([[   0.,  200.],
       [   3.,   -1.],
       [   3.,    5.]])

In [169]:
a.mean(axis=0)

array([  2.,  68.])

In [170]:
a.mean(axis=1)

array([ 100.,    1.,    4.])

In [171]:
a.min(axis=1)

array([ 0., -1.,  3.])

In [172]:
a.max(axis=0)

array([   3.,  200.])

In [173]:
a = np.array([6, 2, 5, -1, 0], float)

In [174]:
sorted(a)

[-1.0, 0.0, 2.0, 5.0, 6.0]

In [175]:
a

array([ 6.,  2.,  5., -1.,  0.])

#### Values in an array can be "clipped" to be within a prespecified range. This is the same as
applying min(max(x, minval), maxval) to each element x in an array. Like winsorization

In [176]:
a = np.array([6, 2, 5, -1, 0], float)
a.clip(0,5)

array([ 5.,  2.,  5.,  0.,  0.])

In [177]:
a = np.array([1, 1, 4, 5, 5, 5, 7], float)
np.unique(a)

array([ 1.,  4.,  5.,  7.])

In [178]:
a = np.array([[1, 2], [3, 4]], float)
print(a)

[[ 1.  2.]
 [ 3.  4.]]


In [179]:
a.diagonal()

array([ 1.,  4.])

In [180]:
a = np.array([1, 3, 0], float)
b = np.array([0, 3, 2], float)
a > b

array([ True, False, False], dtype=bool)

In [181]:
a < b

array([False, False,  True], dtype=bool)

In [182]:
a == b

array([False,  True, False], dtype=bool)

In [183]:
 np.logical_and(a > b, a < b)

array([False, False, False], dtype=bool)

In [184]:
np.logical_or(np.logical_or(a > b, b > a), a == b)

array([ True,  True,  True], dtype=bool)

In [185]:
a = np.array([1, 3, 0], float)
a > 2

array([False,  True, False], dtype=bool)

#### The any and all operators can be used to determine whether or not any or all elements of a
Boolean array are true:

In [186]:
c = np.array([ True, False, False], bool)
any(c)

True

In [187]:
all(c)

False

The *** where ***  function forms a new array from two arrays of equivalent size using a Boolean filter to choose between elements of the two. Its basic syntax is where(boolarray,
truearray, falsearray):

In [188]:
a = np.array([1, 3, 0], float)
np.where(a != 0.0, 1.0/a, a)

/Users/jvsingh/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:2: RuntimeWarning: divide by zero encountered in divide
  from ipykernel import kernelapp as app


array([ 1.        ,  0.33333333,  0.        ])

In [189]:
np.where(a > 0, 3, 2)

array([3, 3, 2])

A number of functions allow testing of the values in an array. The *** nonzero *** function gives a tuple of indices of the nonzero values in an array. The number of items in the tuple equals the number of axes of the array:

In [190]:
a = np.array([[10, 0], [0, 5]], float)
print a
a.nonzero()

[[ 10.   0.]
 [  0.   5.]]


(array([0, 1]), array([0, 1]))

In [191]:
a = np.array([[10, 1], [3, 0]], float)
print a
a.nonzero()

[[ 10.   1.]
 [  3.   0.]]


(array([0, 0, 1]), array([0, 1, 0]))

#### It is also possible to test whether or not values are NaN ("not a number") or finite:

In [192]:
a = np.array([1, np.NaN, np.Inf], float)
a

array([  1.,  nan,  inf])

In [193]:
np.isnan(a)

array([False,  True, False], dtype=bool)

In [194]:
np.isfinite(a)

array([ True, False, False], dtype=bool)

#### Array item selection and manipulation

In [195]:
a = np.array([[6, 4], [5, 9]], float)
print a
a >= 6
print a >= 6

[[ 6.  4.]
 [ 5.  9.]]
[[ True False]
 [False  True]]


In [196]:
sel = (a >= 6)
a[sel]

array([ 6.,  9.])

In [197]:
sel = a > 0
a[sel]

array([ 6.,  4.,  5.,  9.])

In [198]:
a[np.logical_and(a > 5, a < 9)]

array([ 6.])

##### Selection using integer arrays

In [199]:
a = np.array([2, 4, 6, 8], float)
b = np.array([0, 0, 1, 3, 2, 1], int)
a[b]

array([ 2.,  2.,  4.,  8.,  6.,  4.])

##### Lists can also be used as selection arrays:

In [200]:
a = np.array([2, 4, 6, 8], float)
a[[0, 0, 1, 3, 2, 1]]

array([ 2.,  2.,  4.,  8.,  6.,  4.])

#### Multidimensional selection

In [201]:
a = np.array([[1, 4], [9, 16]], float)
print a

[[  1.   4.]
 [  9.  16.]]


In [202]:
b = np.array([0, 0, 1, 1, 0], int)
c = np.array([0, 1, 1, 1, 1], int)
a[b,c] #first index for first array and second from second array

array([  1.,   4.,  16.,  16.,   4.])

In [203]:
# ** take *** function
a = np.array([2, 4, 6, 8], float)
b = np.array([0, 0, 1, 3, 2, 1], int)
a.take(b)

array([ 2.,  2.,  4.,  8.,  6.,  4.])

In [204]:
a = np.array([[0, 1], [2, 3]], float)
a

array([[ 0.,  1.],
       [ 2.,  3.]])

In [205]:
b = np.array([0, 0, 1], int)

In [206]:
a.take(b, axis=0)

array([[ 0.,  1.],
       [ 0.,  1.],
       [ 2.,  3.]])

In [207]:
a.take(b, axis=1)

array([[ 0.,  0.,  1.],
       [ 2.,  2.,  3.]])

*** put *** function (opposite of take)

In [208]:
a = np.array([0, 1, 2, 3, 4, 5], float)
b = np.array([9, 8, 7], float)
a.put([0, 3], b)
a

array([ 9.,  1.,  2.,  8.,  4.,  5.])

#### Vector and matrix mathematics

In [211]:
a = np.array([1, 4, 0], float)
b = np.array([2, 2, 1], float)
np.dot(a,b)

10.0

In [212]:
np.outer(a, b)

array([[ 2.,  2.,  1.],
       [ 8.,  8.,  4.],
       [ 0.,  0.,  0.]])

In [213]:
np.inner(a, b)

10.0

In [214]:
np.cross(a, b)

array([ 4., -1., -6.])

#### Linear Algebra calculations

In [217]:
a = np.array([[4, 2, 0], [9, 3, 7], [1, 2, 1]], float)
print a
np.linalg.det(a)

[[ 4.  2.  0.]
 [ 9.  3.  7.]
 [ 1.  2.  1.]]


-48.000000000000028

In [216]:
np.linalg.det(np.array([[4, 2], [9, 3]], float))

-6.0000000000000027

In [218]:
# eigenvalues and eigenvectors 

np.linalg.eig(a)

(array([ 8.85591316,  1.9391628 , -2.79507597]),
 array([[-0.3663565 , -0.54736745,  0.25928158],
        [-0.88949768,  0.5640176 , -0.88091903],
        [-0.27308752,  0.61828231,  0.39592263]]))

In [220]:
vals, vecs= np.linalg.eig(a)
print vecs
print vals

[[-0.3663565  -0.54736745  0.25928158]
 [-0.88949768  0.5640176  -0.88091903]
 [-0.27308752  0.61828231  0.39592263]]
[ 8.85591316  1.9391628  -2.79507597]


In [226]:
x_1 = vecs[:,0] #first eigenvector
lambda_1 = vals[0] #first eigenvalue
np.dot(a, x_1) - lambda_1 * x_1  

array([ -2.66453526e-15,  -4.44089210e-15,  -1.33226763e-15])

In [228]:
#should be close to 0,0,0
np.dot(a, vecs[:,1]) - vals[1] * vecs[:,1]

array([ -1.11022302e-15,  -3.77475828e-15,   0.00000000e+00])

In [227]:
#should be close to 0,0,0
np.dot(a, vecs[:,2]) - vals[2] * vecs[:,2]

array([  0.00000000e+00,   0.00000000e+00,   2.22044605e-16])

In [229]:
b = np.linalg.inv(a)

In [230]:
b

array([[ 0.22916667,  0.04166667, -0.29166667],
       [ 0.04166667, -0.08333333,  0.58333333],
       [-0.3125    ,  0.125     ,  0.125     ]])

In [231]:
np.dot(a, b)

array([[  1.00000000e+00,   0.00000000e+00,  -2.22044605e-16],
       [  0.00000000e+00,   1.00000000e+00,   0.00000000e+00],
       [  0.00000000e+00,   0.00000000e+00,   1.00000000e+00]])

In [232]:
a = np.array([[1, 3, 4], [5, 2, 3]], float)

In [233]:
U, s, Vh = np.linalg.svd(a)

In [234]:
U

array([[-0.6113829 , -0.79133492],
       [-0.79133492,  0.6113829 ]])

In [235]:
s

array([ 7.46791327,  2.86884495])

#### Polynomial mathematics

##### Polynomial coefficients from set of roots

In [236]:
np.poly([-1, 1, 1, 10])

array([  1., -11.,   9.,  11., -10.])

so polynomial is $ x^4 - 11 x^3 + 9 x^2 + 11x - 10 $

In [238]:
#roots from polynomial
np.roots([1, 4, -2, 3])

array([-4.57974010+0.j        ,  0.28987005+0.75566815j,
        0.28987005-0.75566815j])

Polynomial integration:

$x^3  + x^2 + x + 1$

In [239]:
 np.polyint([1, 1, 1, 1])

array([ 0.25      ,  0.33333333,  0.5       ,  1.        ,  0.        ])

means $ 0.25 x^4 + 0.333 x^3  + 0.5 x^2 + x + C $ where C is set 0

There are other routines for plynomials for polyadd, polysub, polymul, and polydiv. Also to evalutae plynomial at a given point you may use ** polyval **

the *** polyfit *** function can be used to fit a polynomial of specified order to a set of data using a least-squares approach:

In [240]:
x = [1, 2, 3, 4, 5, 6, 7, 8]
y = [0, 2, 1, 3, 7, 10, 11, 19]
np.polyfit(x, y, 2) #returns the plynomial coefficients

array([ 0.375     , -0.88690476,  1.05357143])

### Statistics functions

In [242]:
a = np.array([1, 4, 3, 8, 9, 2, 3], float)
np.median(a)

3.0

In [243]:
a = np.array([1, 4, 3, 8, 9, 2, 3], float)
np.median(a)

3.0

In [245]:
a = np.array([[1, 2, 1, 3], [5, 3, 1, 8]], float)
np.corrcoef(a)

array([[ 1.        ,  0.72870505],
       [ 0.72870505,  1.        ]])

In [246]:
np.cov(a)

array([[ 0.91666667,  2.08333333],
       [ 2.08333333,  8.91666667]])

In [248]:
2.08333333/np.sqrt(0.91666667 * 8.91666667)

0.72870504403430825

### Random numbers